File: agent/internal/collect/listeners.go
package collect

// TCP ports the host is currently listening on.
//
// WHY THIS IS NOT A PROBE. probe.go answers "is the port the server asked about
// up"; this answers "what is here that nobody has registered yet". The point is
// to spare somebody an ssh and an `lsof` before they can register a service —
// the dev server they started a minute ago is already in this list.
//
// ⚠ THIS COLLECTOR IS CACHED, AND THE CACHE IS THE POINT. Reading it costs a
// spawned process on darwin, and on linux naming the process behind a port means
// walking /proc/<pid>/fd. Paying either every heartbeat is precisely the cost
// pattern the usage track was rewritten to remove (docs/USAGE-COLLECTION.md §1:
// the collection must not cost more than the agent it measures). Ports do not
// move on a five-second scale, so one answer is reused for ListenerTTLSec.
//
// ⚠ NOTHING HERE GOES NEAR THE `exec` FRAME. That path exists so a person can
// run a command through MCP — it needs write scope, caps output at 64 KB and
// admits four at a time. A collector that runs on a timer is not its caller.

import (
	"context"
	"net"
	"os"
	"path/filepath"
	"runtime"
	"slices"
	"strconv"
	"strings"
	"sync"

	"github.com/podosoft-dev/pdmux/agent/internal/protocol"
	"github.com/podosoft-dev/pdmux/agent/internal/sys"
)

// MaxListeners is heartbeat.listeners' maxItems in the contract.
//
// ⚠ EXCEEDING IT REJECTS THE WHOLE HEARTBEAT rather than the extra rows — the
// resource bars and the session list would go down with the frame. Truncating
// here is what keeps a host with an unusual number of open ports visible at all,
// and Dropped is what stops the truncation from reading as the whole truth.
const MaxListeners = 128

// listenerProcessMax is listener.process's maxLength in the contract.
const listenerProcessMax = 64

// ListenerTTLSec is how long one reading is reused. See the header.
const ListenerTTLSec = 60

// defaultListenerTimeoutMs bounds the darwin lookup. A wedged `lsof` costs this
// pass its port list and nothing else.
const defaultListenerTimeoutMs = 4_000

// procFDBudget caps the linux name pass. Resolving a port's process means
// readlink() on every file descriptor of every process until the socket inode
// turns up, and on a busy host that is tens of thousands of syscalls. The budget
// makes the worst case bounded rather than proportional to how loaded the host
// is; ports whose name is not reached keep their row and lose only the name.
const procFDBudget = 20_000

// ListenerReading is the port list plus the two facts that decide how it renders.
type ListenerReading struct {
	Listeners []protocol.Listener
	// Supported is whether this host could be asked at all. False means the tool
	// needed to look is absent — which is NOT the claim "nothing is listening",
	// and an empty list must never be shown as if it were.
	Supported bool
	// Dropped counts ports the contract's cap left out.
	Dropped int
}

// resolveLsof names the binary to run on darwin.
//
// ⚠ A VARIABLE FOR THE SAME REASON resolveMux IS ONE. Two of the directories
// ResolveBinary searches are absolute system prefixes, so a spec that clears
// PATH still finds the machine's own lsof and passes by measuring the host
// instead of the case it meant to state. That false pass has already happened
// once in this package.
var resolveLsof = func() string { return sys.ResolveBinary("lsof", "") }

// procNetRoot is where the linux reader looks. A variable so a spec can point it
// at a fixture directory and exercise the real reader rather than a parser.
var procNetRoot = "/proc"

// ReadListeners takes one reading, using whatever this OS offers.
func ReadListeners(ctx context.Context, timeoutMs int) ListenerReading {
	switch runtime.GOOS {
	case "linux":
		return readListenersProc()
	case "darwin":
		return readListenersLsof(ctx, timeoutMs)
	}
	// An OS with neither path is not a failure to report every beat; it is a host
	// that cannot answer this question. Supported says so.
	return ListenerReading{Listeners: []protocol.Listener{}, Supported: false}
}

// NewCachedListeners wraps a reader in the TTL the header explains.
//
// The previous answer is served while the TTL holds, including a `Supported:
// false` one — a host without the tool does not grow it between two beats, and
// re-deciding that every five seconds is the cost this cache exists to avoid.
func NewCachedListeners(read func(context.Context) ListenerReading, ttlSec int, now func() int64) func(context.Context) ListenerReading {
	if ttlSec <= 0 {
		ttlSec = ListenerTTLSec
	}
	if now == nil {
		now = nowSeconds
	}
	var mu sync.Mutex
	var cached ListenerReading
	var readAt int64

	return func(ctx context.Context) ListenerReading {
		mu.Lock()
		defer mu.Unlock()
		if readAt != 0 && now()-readAt < int64(ttlSec) {
			return cached
		}
		cached = read(ctx)
		readAt = now()
		return cached
	}
}

// ---------------------------------------------------------------------------
// darwin — lsof
// ---------------------------------------------------------------------------

func readListenersLsof(ctx context.Context, timeoutMs int) ListenerReading {
	// -n and -P keep it from resolving names (a DNS lookup per socket); -sTCP:LISTEN
	// is the filter; -F cn asks for field output and `+c 0` turns off the command
	// truncation. See parseLsofFields for why the last two are not optional.
	result := sys.Run(ctx, resolveLsof(),
		[]string{"-nP", "-iTCP", "-sTCP:LISTEN", "-F", "cn", "+c", "0"},
		sys.Options{TimeoutMs: boundMs(timeoutMs, defaultListenerTimeoutMs)})
	if result.NotFound {
		return ListenerReading{Listeners: []protocol.Listener{}, Supported: false}
	}
	// ⚠ EXIT 1 IS THE ORDINARY ANSWER FOR "NOTHING MATCHED". lsof reports it when
	// its filters select no file, so treating a non-zero exit as a failure would
	// call a quiet host unsupported. Only stdout decides.
	listeners, dropped := parseLsofFields(result.Stdout).result()
	return ListenerReading{Listeners: listeners, Supported: true, Dropped: dropped}
}

// parseLsofFields reads lsof's `-F` field output: one letter per line, and the
// process lines (`p`, `c`) apply to the file lines (`n`) that follow.
//
// ⚠ FIELD OUTPUT RATHER THAN THE TABLE, for two reasons measured on a real host.
// The table truncates COMMAND to nine characters, so `ControlCenter` arrives as
// `ControlCe` and the row names a process that does not exist; and a command
// containing a space — `Google Chrome` — shifts every later column, so splitting
// on whitespace reads the address out of the wrong field. `+c 0` removes the
// truncation and `-F` removes the columns.
func parseLsofFields(text string) *listenerAccumulator {
	acc := newListenerAccumulator()
	process := ""
	for _, raw := range strings.Split(text, "\n") {
		line := strings.TrimRight(raw, "\r")
		if line == "" {
			continue
		}
		switch line[0] {
		case 'p':
			// A new process section. The command arrives on its own line, so the
			// previous process's name must not survive into this one — a `p` with
			// no `c` after it is a nameless row, not a mislabelled one.
			process = ""
		case 'c':
			process = line[1:]
		case 'n':
			if port, loopback, ok := parseListenAddress(line[1:]); ok {
				acc.add(port, process, loopback)
			}
		}
	}
	return acc
}

// parseListenAddress reads the address forms lsof prints: `*:5432`,
// `127.0.0.1:5432` and `[::1]:5432`.
func parseListenAddress(value string) (port int, loopback bool, ok bool) {
	address := strings.TrimSpace(value)
	// `-sTCP:LISTEN` already filtered, but lsof still appends `(LISTEN)` in some
	// versions. Anything after a space is not part of the address.
	if index := strings.IndexByte(address, ' '); index >= 0 {
		address = address[:index]
	}
	index := strings.LastIndexByte(address, ':')
	if index < 0 {
		return 0, false, false
	}
	port, err := strconv.Atoi(address[index+1:])
	if err != nil || port < 1 || port > 65535 {
		return 0, false, false
	}
	return port, isLoopbackHost(strings.Trim(address[:index], "[]")), true
}

// isLoopbackHost decides the flag the whole forwarding question later turns on.
//
// `*` is every interface, which is the opposite of loopback — reporting it as
// loopback would understate what is already reachable from off the host.
func isLoopbackHost(host string) bool {
	if host == "" || host == "*" {
		return false
	}
	ip := net.ParseIP(host)
	return ip != nil && ip.IsLoopback()
}

// ---------------------------------------------------------------------------
// linux — /proc
// ---------------------------------------------------------------------------

// procTCPListen is TCP_LISTEN in the state column of /proc/net/tcp.
const procTCPListen = "0A"

func readListenersProc() ListenerReading {
	acc := newListenerAccumulator()
	inodes := map[uint64]int{}
	found := false
	// Both families, because a dev server bound to ::1 is invisible in the first
	// file and a host that only reads one under-reports without saying so.
	for _, name := range []string{"net/tcp", "net/tcp6"} {
		raw, err := os.ReadFile(filepath.Join(procNetRoot, name))
		if err != nil {
			continue
		}
		found = true
		parseProcNetTCP(string(raw), acc, inodes)
	}
	if !found {
		// No /proc at all. Reporting an empty list here would claim a quiet host.
		return ListenerReading{Listeners: []protocol.Listener{}, Supported: false}
	}
	resolveProcNames(inodes, acc)
	listeners, dropped := acc.result()
	return ListenerReading{Listeners: listeners, Supported: true, Dropped: dropped}
}

// parseProcNetTCP reads the listening rows and records each socket's inode, which
// is the only handle /proc/net/tcp gives towards a process name.
func parseProcNetTCP(text string, acc *listenerAccumulator, inodes map[uint64]int) {
	for _, line := range strings.Split(text, "\n") {
		fields := strings.Fields(line)
		// sl, local, remote, st, tx:rx, tr:when, retrnsmt, uid, timeout, inode
		if len(fields) < 10 || fields[3] != procTCPListen {
			continue
		}
		port, loopback, ok := parseProcAddress(fields[1])
		if !ok {
			continue
		}
		acc.add(port, "", loopback)
		if inode, err := strconv.ParseUint(fields[9], 10, 64); err == nil && inode != 0 {
			inodes[inode] = port
		}
	}
}

// parseProcAddress reads `0100007F:1F90` — the address in hex, the port in hex.
func parseProcAddress(value string) (port int, loopback bool, ok bool) {
	index := strings.LastIndexByte(value, ':')
	if index < 0 {
		return 0, false, false
	}
	parsed, err := strconv.ParseUint(value[index+1:], 16, 32)
	if err != nil || parsed < 1 || parsed > 65535 {
		return 0, false, false
	}
	ip, ok := parseProcIP(value[:index])
	if !ok {
		return 0, false, false
	}
	return int(parsed), ip.IsLoopback(), true
}

// parseProcIP decodes the hex address.
//
// ⚠ IT IS NOT PLAIN BIG-ENDIAN HEX, and reading it as such gets loopback exactly
// backwards. The kernel prints each 32-bit word in HOST byte order, so on a
// little-endian machine `0100007F` is 127.0.0.1 and not 1.0.0.127. IPv6 is the
// same rule applied to four words, which is why ::1 appears as twenty-four zeros
// followed by `01000000`.
func parseProcIP(value string) (net.IP, bool) {
	if len(value) == 0 || len(value)%8 != 0 {
		return nil, false
	}
	raw := make([]byte, 0, len(value)/2)
	for start := 0; start < len(value); start += 8 {
		word, err := strconv.ParseUint(value[start:start+8], 16, 32)
		if err != nil {
			return nil, false
		}
		raw = append(raw, byte(word), byte(word>>8), byte(word>>16), byte(word>>24))
	}
	if len(raw) != net.IPv4len && len(raw) != net.IPv6len {
		return nil, false
	}
	return net.IP(raw), true
}

// resolveProcNames walks /proc/<pid>/fd looking for the socket inodes the port
// list needs a name for.
//
// ⚠ THE NAME COMES FROM /proc/<pid>/comm, NEVER FROM cmdline. A command line
// carries arguments, and this repository's own installer takes a single-use
// enrollment code as one — putting argv on the wire would send exactly the thing
// the credential rule keeps out of logs and unit files.
//
// Unreadable processes are skipped rather than reported: an agent running as a
// user cannot see root's descriptors, and a port with no name is still a useful
// row.
func resolveProcNames(inodes map[uint64]int, acc *listenerAccumulator) {
	if len(inodes) == 0 {
		return
	}
	entries, err := os.ReadDir(procNetRoot)
	if err != nil {
		return
	}
	remaining := len(inodes)
	budget := procFDBudget
	for _, entry := range entries {
		if remaining == 0 || budget <= 0 {
			return
		}
		pid := entry.Name()
		if _, err := strconv.Atoi(pid); err != nil {
			continue
		}
		fdDir := filepath.Join(procNetRoot, pid, "fd")
		fds, err := os.ReadDir(fdDir)
		if err != nil {
			continue
		}
		name := ""
		for _, fd := range fds {
			if budget <= 0 {
				return
			}
			budget--
			link, err := os.Readlink(filepath.Join(fdDir, fd.Name()))
			if err != nil {
				continue
			}
			inode, ok := socketInode(link)
			if !ok {
				continue
			}
			port, wanted := inodes[inode]
			if !wanted {
				continue
			}
			if name == "" {
				name = procComm(pid)
			}
			acc.name(port, name)
			delete(inodes, inode)
			remaining--
		}
	}
}

// socketInode reads the `socket:[12345]` form a socket descriptor links to.
func socketInode(link string) (uint64, bool) {
	const prefix = "socket:["
	if !strings.HasPrefix(link, prefix) || !strings.HasSuffix(link, "]") {
		return 0, false
	}
	inode, err := strconv.ParseUint(link[len(prefix):len(link)-1], 10, 64)
	if err != nil {
		return 0, false
	}
	return inode, true
}

func procComm(pid string) string {
	raw, err := os.ReadFile(filepath.Join(procNetRoot, pid, "comm"))
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(raw))
}

// ---------------------------------------------------------------------------
// Merging what the two readers produce
// ---------------------------------------------------------------------------

type listenerEntry struct {
	process  string
	loopback bool
}

// listenerAccumulator merges the bindings of one port into one row. A port bound
// on several addresses is one thing to a person, and listing 5432 three times
// makes the table unreadable on any host with dual-stack services.
type listenerAccumulator struct {
	byPort map[int]*listenerEntry
}

func newListenerAccumulator() *listenerAccumulator {
	return &listenerAccumulator{byPort: map[int]*listenerEntry{}}
}

func (a *listenerAccumulator) add(port int, process string, loopback bool) {
	if port < 1 || port > 65535 {
		return
	}
	entry, seen := a.byPort[port]
	if !seen {
		a.byPort[port] = &listenerEntry{process: process, loopback: loopback}
		return
	}
	// ⚠ ONE NON-LOOPBACK BINDING SETTLES IT. A port listening on both 127.0.0.1
	// and 0.0.0.0 is reachable from off the host, and the safe direction for this
	// flag is the one that does not understate exposure.
	entry.loopback = entry.loopback && loopback
	if entry.process == "" {
		entry.process = process
	}
}

func (a *listenerAccumulator) name(port int, process string) {
	if entry, seen := a.byPort[port]; seen && entry.process == "" {
		entry.process = process
	}
}

// result sorts by port and applies the contract's cap.
//
// Sorted because an unordered table reshuffles between beats and reads as
// activity; by port because that is what somebody scanning the list is looking
// for.
func (a *listenerAccumulator) result() ([]protocol.Listener, int) {
	ports := make([]int, 0, len(a.byPort))
	for port := range a.byPort {
		ports = append(ports, port)
	}
	slices.Sort(ports)

	dropped := 0
	if len(ports) > MaxListeners {
		dropped = len(ports) - MaxListeners
		ports = ports[:MaxListeners]
	}

	out := make([]protocol.Listener, 0, len(ports))
	for _, port := range ports {
		entry := a.byPort[port]
		item := protocol.NewListener()
		item.Port = port
		item.Process = truncateRunes(entry.process, listenerProcessMax)
		item.LoopbackOnly = entry.loopback
		out = append(out, item)
	}
	return out, dropped
}

File: packages/core/src/listeners.ts
/**
 * Discovered ports: what the host is listening on that nobody has registered.
 *
 * WHY THIS IS A JUDGEMENT AND NOT A FILTER. A developer's machine listens on far
 * more ports than it runs services — measured on one Mac: fifty-nine, of which
 * maybe a dozen were anything a person had started on purpose. Printing all
 * fifty-nine is the same as printing none.
 *
 * ⚠ BUT NOTHING IS DELETED, ONLY FOLDED, and every fold carries the reason it was
 * folded. A list that silently drops the port somebody is looking for does not
 * read as "we filtered that"; it reads as "this tool is broken", and the person
 * goes back to ssh and lsof — which is the thing this feature exists to replace.
 * Guessing wrong is acceptable. Guessing invisibly is not.
 */

export interface ListenerInput {
	port: number;
	process?: string | null;
	loopbackOnly?: boolean | null;
}

/** Why a port was folded away from the default view. `null` means it is shown. */
export type FoldReason = 'ephemeral' | 'system-port' | 'system-process';

export interface DiscoveredPort {
	port: number;
	/** Executable name, or '' when the host could not attribute the socket. */
	process: string;
	loopbackOnly: boolean;
	folded: FoldReason | null;
}

export interface DiscoveredPorts {
	shown: DiscoveredPort[];
	folded: DiscoveredPort[];
}

/**
 * Start of the ephemeral range (IANA, and the default on macOS).
 *
 * ⚠ THIS IS THE MOST USEFUL RULE HERE, and it is not about what the port is — it
 * is about whether it will still be that port tomorrow. A listener the kernel
 * handed out is different after every restart, so registering it as a service
 * produces a row that is wrong by the next reboot. Twenty of the fifty-nine ports
 * measured were this.
 */
export const EPHEMERAL_PORT_MIN = 49152;

/**
 * Ports that belong to the operating system or to the machine's own services.
 *
 * Deliberately short. Every entry folds a port for EVERY host, so a wrong guess
 * here is wrong everywhere — the bar is "no developer ever starts this on
 * purpose", not "this is usually infrastructure". Postgres and Redis are not
 * here for exactly that reason: on a developer's machine they are the work.
 */
export const SYSTEM_PORTS: ReadonlySet<number> = new Set([
	22, // ssh
	25, // smtp
	53, // dns
	88, // kerberos
	111, // rpcbind
	123, // ntp
	137, 138, 139, // netbios
	445, // smb
	548, // afp
	631, // ipp/cups
	3283, // apple remote desktop
	5353, // mdns
	5900, // vnc
	// macOS hands 5000 and 7000 to ControlCenter for AirPlay. They look exactly
	// like a dev server, and on this machine they were the two most confusing
	// rows in the list.
	5000, 7000,
]);

/**
 * Processes whose listening ports are never somebody's service.
 *
 * Matched case-insensitively on the whole executable name — never as a substring,
 * because `node` appears inside plenty of names a developer does own.
 *
 * ⚠ A BROWSER BELONGS HERE. Chrome's remote-debugging port is not infrastructure
 * and not a dev server, but forwarding it hands over every open tab, every cookie
 * and anything being typed into a password field. It is folded by name rather
 * than by port because the port is configurable and the danger is not.
 */
export const SYSTEM_PROCESSES: ReadonlySet<string> = new Set([
	// macOS
	'controlcenter',
	'rapportd',
	'sharingd',
	'airplayxpchelper',
	'remoted',
	'mdnsresponder',
	'netbiosd',
	'identityservicesd',
	'launchd',
	// linux
	'systemd',
	'systemd-resolve',
	'systemd-resolved',
	'sshd',
	'cupsd',
	'avahi-daemon',
	'chronyd',
	'rpcbind',
	'dnsmasq',
	'networkmanager',
	// browsers — see the warning above
	'google chrome',
	'google chrome helper',
	'chromium',
	'firefox',
	'safari',
	'msedge',
	'microsoft edge',
]);

function normalisePort(value: unknown): number | null {
	const port = Math.trunc(Number(value));
	return Number.isFinite(port) && port >= 1 && port <= 65535 ? port : null;
}

/** Why this port is folded, or `null` to show it. */
export function foldReason(entry: ListenerInput): FoldReason | null {
	const port = normalisePort(entry?.port);
	if (port === null) return null;
	if (SYSTEM_PORTS.has(port)) return 'system-port';
	if (SYSTEM_PROCESSES.has(String(entry?.process ?? '').trim().toLowerCase())) return 'system-process';
	// Checked last so a named system process on a high port reads as what it is
	// rather than as "the kernel picked this".
	if (port >= EPHEMERAL_PORT_MIN) return 'ephemeral';
	return null;
}

/**
 * Split the discovered ports into the ones worth showing and the ones to fold.
 *
 * ⚠ PORTS ALREADY REGISTERED AS SERVICES ARE DROPPED FROM BOTH. They are on the
 * services table directly above this one, with a probe result and a link — which
 * is strictly more than this table can say about them. Listing them twice makes
 * the shorter list look like the authoritative one. What is left is therefore
 * exactly the question this section answers: what is running here that pdmux does
 * not know about yet.
 */
export function discoveredPorts(
	listeners: readonly ListenerInput[] | null | undefined,
	registeredPorts: Iterable<number> | null | undefined = [],
): DiscoveredPorts {
	const registered = new Set<number>();
	for (const value of registeredPorts ?? []) {
		const port = normalisePort(value);
		if (port !== null) registered.add(port);
	}

	const seen = new Set<number>();
	const shown: DiscoveredPort[] = [];
	const folded: DiscoveredPort[] = [];

	for (const entry of Array.isArray(listeners) ? listeners : []) {
		const port = normalisePort(entry?.port);
		if (port === null || registered.has(port) || seen.has(port)) continue;
		seen.add(port);

		const reason = foldReason(entry);
		const item: DiscoveredPort = {
			port,
			process: String(entry?.process ?? '').trim(),
			// Absent means unknown, and the safe reading of unknown exposure is the
			// one that does not promise the port is confined to the host.
			loopbackOnly: entry?.loopbackOnly === true,
			folded: reason,
		};
		(reason === null ? shown : folded).push(item);
	}

	shown.sort((a, b) => a.port - b.port);
	folded.sort((a, b) => a.port - b.port);
	return { shown, folded };
}

/**
 * Narrow the list to what a person typed.
 *
 * Matches a port by the digits it contains rather than by prefix — on a machine
 * running four databases the useful query is "543", and a prefix match would
 * find 5432 but not 15432, which is the one they are actually looking for.
 * Process names match case-insensitively on a substring for the same reason.
 *
 * ⚠ THE CALLER MUST SEARCH THE FOLDED PORTS TOO. Folding is a guess about what
 * is worth showing FIRST; it is not a claim that those ports do not exist. A
 * search that only looks at the visible half answers "no results" for a port
 * that is right there, which is precisely the "this tool is broken" reaction the
 * fold was designed to avoid — and the person goes back to ssh. This function
 * filters whatever it is given, so give it everything.
 */
export function filterPorts(
	ports: readonly DiscoveredPort[] | null | undefined,
	query: unknown,
): DiscoveredPort[] {
	const rows = Array.isArray(ports) ? ports.filter(Boolean) : [];
	const needle = String(query ?? '').trim().toLowerCase();
	if (!needle) return [...rows];
	return rows.filter(
		(row) => String(row.port).includes(needle) || row.process.toLowerCase().includes(needle),
	);
}

/** Longest label the API accepts for a service. */
const LABEL_MAX = 64;

/**
 * A label for promoting a discovered port into a service.
 *
 * The port is part of the name rather than a fallback: a host running four
 * postgres instances is the ordinary case on a developer's machine, and
 * `postgres` alone would collide with the second one. Labels are unique per host
 * and the API rejects a duplicate, so the suffix loop is what keeps the button
 * from opening a dialog that cannot be submitted.
 */
export function suggestServiceLabel(
	entry: ListenerInput,
	taken: Iterable<string> | null | undefined = [],
): string {
	const port = normalisePort(entry?.port);
	const slug = String(entry?.process ?? '')
		.trim()
		.toLowerCase()
		.replace(/[^a-z0-9]+/g, '-')
		.replace(/^-+|-+$/g, '');

	const base = (slug ? `${slug}-${port ?? 0}` : `port-${port ?? 0}`).slice(0, LABEL_MAX);
	const used = new Set<string>();
	for (const value of taken ?? []) used.add(String(value ?? '').trim().toLowerCase());
	if (!used.has(base.toLowerCase())) return base;

	for (let suffix = 2; suffix < 100; suffix += 1) {
		const tail = `-${suffix}`;
		const candidate = `${base.slice(0, LABEL_MAX - tail.length)}${tail}`;
		if (!used.has(candidate.toLowerCase())) return candidate;
	}
	return base;
}
